drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1181 @@
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "SRC_X", 0) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "SRC_Y", 0) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "SRC_W", buf->width << 16) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "SRC_H", buf->height << 16) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_X", 0) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_Y", 0) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_W", buf->width) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_H", buf->height) < 0)
+		return -1;
+
+	return 0;
+}
+
+/*
+ * A short helper function to compute a changing color value. No need to
+ * understand it.
+ */
+
+static uint8_t next_color(bool *up, uint8_t cur, unsigned int mod)
+{
+	uint8_t next;
+
+	next = cur + (*up ? 1 : -1) * (rand() % mod);
+	if ((*up && next < cur) || (!*up && next > cur)) {
+		*up = !*up;
+		next = cur;
+	}
+
+	return next;
+}
+
+/*
+ * Draw on back framebuffer before the page-flip is requested.
+ */
+
+static void modeset_paint_framebuffer(struct modeset_output *out)
+{
+	struct modeset_buf *buf;
+	unsigned int j, k, off;
+
+	/* draw on back framebuffer */
+	out->r = next_color(&out->r_up, out->r, 5);
+	out->g = next_color(&out->g_up, out->g, 5);
+	out->b = next_color(&out->b_up, out->b, 5);
+	buf = &out->bufs[out->front_buf ^ 1];
+	for (j = 0; j < buf->height; ++j) {
+		for (k = 0; k < buf->width; ++k) {
+			off = buf->stride * j + k * 4;
+			*(uint32_t*)&buf->map[off] =
+				     (out->r << 16) | (out->g << 8) | out->b;
+		}
+	}
+
+}
+
+/*
+ * modeset_draw_out() prepares the framebuffer with the drawing and then it asks
+ * for the driver to perform an atomic commit. This will lead to a page-flip and
+ * the content of the framebuffer will be displayed. In this simple example
+ * we're only using the primary plane, but we could also be updating other
+ * planes in the same atomic commit.
+ *
+ * Just like in modeset_perform_modeset(), we first setup everything with
+ * modeset_atomic_prepare_commit() and then actually perform the atomic commit.
+ * The main difference is that here we have already painted the framebuffer and
+ * also we don't use the flag DRM_MODE_ALLOW_MODESET anymore, since the modeset
+ * already happened. If we continue using this flag, it may result in unecessary
+ * latency which is bad since we want to display content in a predictable and
+ * reasonable rate.
+ */
+
+static void modeset_draw_out(int fd, struct modeset_output *out)
+{
+	drmModeAtomicReq *req;
+	int ret, flags;
+	struct modeset_output *iter;
+
+	/* draw on framebuffer of the output */
+	modeset_paint_framebuffer(out);
+
+	/* prepare all outputs for atomic commit */
+	req = drmModeAtomicAlloc();
+	for (iter = output_list; iter; iter = iter->next) {
```

Comment:
This should run the one output only, not all outputs.

Can you see why?

Think about having two outputs with different refresh rates. Attempting to pageflip on a CRTC that has not yet completed the previous pageflip will fail with `EBUSY`.

Reply:
It makes sense to me now. The event happened because a specific output is able to perform a page-flip again, but we don't know about the state of the others. They may be in the middle of a scanout, for instance.

Reply:
Exactly.